new_lessons/sets.md

```diff
@@ -0,0 +1,337 @@
+
+```python
+
+list_a = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+list_b = [6, 7, 8, 9, 10, 11, 12, 13, 14]
+
+result_list = []
+for element_a in list_a:
+    if element_a in list_b:
+        result_list.append(element_a)
+
+return result_list
+```
+
+This solution will get the job done, but it's not very efficient or elegant!  We can use sets to simplify this code by converting our lists into sets and then using set functionality to solve the problem.
+
+### Creating Sets From Other Data Structures
+
+First we need to turn our lists into sets.  The syntax for this is:
+
+```python
+
+list_a = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+list_b = [6, 7, 8, 9, 10, 11, 12, 13, 14]
+
+set_a = set(list_a) # converts the list into a set!
+set_b = set(list_a) 
+
+```
+
+Next, we can use set intersection to generate a set of all of the elements that are contained in both sets:
+
+```python
+
+result_set = set_a & set b
+
+```
+
+If we want the results in a list we can convert a set into a list in a similar way to how we converted the list into a set:
+
+```python
+
+result_list = list(result_set)
+
+```
+
+Here's the final version of the code:
+
+```python
+
+list_a = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+list_b = [6, 7, 8, 9, 10, 11, 12, 13, 14]
+
+set_a = set(list_a) # converts the list into a set!
+set_b = set(list_a) 
+
+result_set = set_a & set b
+
+return list(result_set)
+
+```
+
+### Iterating Through Sets
+
+Let's say we have a string of letters and we want to find all of the unique characters in the string.  Here's a brute force solution that we could use:
+
+```python
+
+s = "she sells sea shells by the sea shore"
+result = []
+
+for char in s:  # loop over each character in the string
+    if char not in result: # if the character is not in the result set:
+        result.append(char) # add the character to the result set
+
+for char in result: #print out the results
+    print(char)
+
+```
+
+Just like our other brute force solution, this will get the job done, but we can use sets to come up with a more streamlined solution!
+
+```python
+
+s = "she sells sea shells by the sea shore"
+result_set = set(s) # that's it!  the set now contains {'o', 'a', 'b', 'y', 'r', 't', 'l', 's', 'e', 'h', ' '}
+
+```
+
+If we want to print out the set, we can use a for loop just like the for loop we used for the list above:
+
+```python
+
+for char in result_set:
+    print(char)
+
+```
+
+Here's our final solution:
+
+```python
+
+s = "she sells sea shells by the sea shore"
+result_set = set(s)
+for char in result_set:
+    print(char)
+
+```
+
```

Comment:
```suggestion

## Check for Understanding

```

Reply:
I know there's likely to be a problem set attached to sets, but I like problem sets to be... problem sets, and questions at the end of lessons in "Check for Understanding" to be on the very fundamental layer. I'd personally recommend adding MC questions to reinforce:

- How to make a set in Python
- The uniqueness of elements in a set
- What intersection does
- What union does